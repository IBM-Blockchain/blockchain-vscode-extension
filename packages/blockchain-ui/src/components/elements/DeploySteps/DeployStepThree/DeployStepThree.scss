html{
    overflow: overlay !important;
    
    #root {
        .deploy-page-container {
            #customize{
                p {
                    margin-bottom: 8px;
                }
            }
        }
    }
}