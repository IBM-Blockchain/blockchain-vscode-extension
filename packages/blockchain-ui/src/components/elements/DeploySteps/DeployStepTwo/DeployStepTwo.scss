.deploy-step-two-notification {
    .bx--inline-notification__text-wrapper {
        flex-direction: column;
        align-items: flex-start !important;
    }
}
