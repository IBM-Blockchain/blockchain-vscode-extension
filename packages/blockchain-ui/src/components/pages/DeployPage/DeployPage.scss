@import '../../../mixins/layout-mixins.scss';
@import '@carbon/colors/scss/colors';
@import '@carbon/layout/scss/_spacing';
@import '@carbon/grid/scss/grid';
@import "carbon-components/scss/globals/scss/_css--helpers";


html{
    overflow: overlay !important;
    
    #root {
        .deploy-page-container {
            
    
            .progress-row{
                margin-top: $spacing-05;
                margin-bottom: $spacing-08;
            }
    
            .dropdown-row{
                margin-bottom: $spacing-09 * 2;
            }
    
            .margin-bottom-06 {
                margin-bottom: $spacing-06;
    
                
            }
    
            .margin-bottom-05 {
                margin-bottom: $spacing-05;
            }
    
            .margin-right-06{
                margin-right: $spacing-06;
            }        
    
            .heading-combo-container{
                p {
                    font-size: 0.875rem;
                }
            }
    
            #buttonList {
                display: flex;
                justify-content: flex-end;
                height: fit-content;
            }
        }
    
    }
}
