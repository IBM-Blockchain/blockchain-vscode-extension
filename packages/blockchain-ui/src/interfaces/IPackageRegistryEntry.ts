interface IPackageRegistryEntry {
    name: string;
    path: string;
    version?: string;
    sizeKB: number;
}

export default IPackageRegistryEntry;
