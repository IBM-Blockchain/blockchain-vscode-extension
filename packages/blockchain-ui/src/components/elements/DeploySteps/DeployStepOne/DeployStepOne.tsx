import React, { Component } from 'react';
import { Dropdown, Accordion, AccordionItem, InlineNotification, NotificationActionButton, Link, Tag } from 'carbon-components-react';
import IPackageRegistryEntry from '../../../../interfaces/IPackageRegistryEntry';
import ThemedImage from '../../../elements/ThemedImage/ThemedImage';
import lightStep1 from '../../../../resources/lightStep1.svg';
import lightStep2 from '../../../../resources/lightStep2.svg';
import lightStep3 from '../../../../resources/lightStep3.svg';
import lightStep4 from '../../../../resources/lightStep4.svg';
import darkStep1 from '../../../../resources/darkStep1.svg';
import darkStep2 from '../../../../resources/darkStep2.svg';
import darkStep3 from '../../../../resources/darkStep3.svg';
import darkStep4 from '../../../../resources/darkStep4.svg';
import './DeployStepOne.scss';

interface IProps {
    workspaceNames: string[];
    deletedSelectedPackage: boolean;
    packageEntries: IPackageRegistryEntry[];
    selectedWorkspace: string | undefined;
    selectedPackage: IPackageRegistryEntry | undefined; // Package the user has selected in dropdown
    onPackageChange: (selectedPackage: IPackageRegistryEntry | undefined, workspaceName?: string) => void; // Callback to DeployPage
    onPackageWorkspace: (workspaceName: string) => void;
}

interface StepOneState {
    workspaceNames: string[];
    packageEntries: IPackageRegistryEntry[];
}

class DeployStepOne extends Component<IProps, StepOneState> {

    unselectedPackageLabel: string = 'Select smart contract';

    constructor(props: Readonly<IProps>) {
        super(props);
        this.state = {
            packageEntries: this.props.packageEntries,
            workspaceNames: this.props.workspaceNames.length > 0 ? this.props.workspaceNames : []
        };

        this.formatAllEntries = this.formatAllEntries.bind(this);
        this.formatPackageEntry = this.formatPackageEntry.bind(this);
        this.formatWorkspaceEntry = this.formatWorkspaceEntry.bind(this);
        this.selectPackage = this.selectPackage.bind(this);
        this.packageWorkspace = this.packageWorkspace.bind(this);

    }

    componentWillReceiveProps(receivedProps: IProps): void {

        let receivedNewPackages: boolean = false;

        if (receivedProps.packageEntries.length === this.state.packageEntries.length) {
            // Check each individual entry
            for (let i: number = 0; i < receivedProps.packageEntries.length; i++) {
                const newProp: IPackageRegistryEntry = receivedProps.packageEntries[i];
                const currentState: IPackageRegistryEntry = this.state.packageEntries[i];
                if (newProp.name !== currentState.name || newProp.version !== currentState.version) {
                    receivedNewPackages = true;
                    break;
                }
            }
        } else {
            // Different length, so different packages
            receivedNewPackages = true;
        }

        if (receivedNewPackages) {
            this.setState({packageEntries: receivedProps.packageEntries});
        }
    }

    selectPackage(data: any): void {

        // TODO: At some point we should move PackageRegistry to common, and call 'get' from here.
        // There would also be no need to pass in all entries when creating DeployView.

        let packageName: string;
        let packageVersion: string | undefined;

        if (data.selectedItem.includes(' (packaged)')) {
            // If selected a package

             // Get 'name@version' or just 'name'
            const packageIdentifier: string = data.selectedItem.split(' (packaged')[0];
            if (packageIdentifier.includes('@')) {
                // If it has a version
                const _splitParts: string[] = packageIdentifier.split('@');
                packageName = _splitParts[0];
                packageVersion = _splitParts[1];
            } else {
                // If it doesn't have a version
                packageName = packageIdentifier;
            }

            const selectedPackage: IPackageRegistryEntry | undefined = this.state.packageEntries.find((_entry: IPackageRegistryEntry) => {
                return (packageVersion && _entry.name === packageName && _entry.version === packageVersion) || (!packageVersion && _entry.name === packageName);
            }) as IPackageRegistryEntry;

            this.props.onPackageChange(selectedPackage);
        } else {
            // If selected an open workspace
            this.props.onPackageChange(undefined, data.selectedItem.split(' (open project)')[0]);

        }

    }

    formatAllEntries(): string[] {
        const packageNames: string[] = [];
        for (const name of this.state.workspaceNames) {
            const entryName: string = this.formatWorkspaceEntry(name);
            packageNames.push(entryName);
        }

        for (const entry of this.state.packageEntries) {
            const entryName: string = this.formatPackageEntry(entry);
            packageNames.push(entryName);
        }

        return packageNames;
    }

    formatPackageEntry(entry: IPackageRegistryEntry): string {
        if (entry.version) {
            return `${entry.name}@${entry.version} (packaged)`;
        } else {
            return `${entry.name} (packaged)`;
        }
    }

    formatWorkspaceEntry(workspaceName: string): string {
        return `${workspaceName} (open project)`;
    }

    packageWorkspace(): void {
        this.props.onPackageWorkspace(this.props.selectedWorkspace as string);
    }

    render(): JSX.Element {
        const items: string[] = this.formatAllEntries();

        let deletedMessage: JSX.Element | undefined;
        let packageMessage: JSX.Element | undefined;
        let packageStillExists: boolean = false;

        if (this.props.selectedPackage) {

            const packageToCheck: IPackageRegistryEntry | undefined = this.state.packageEntries.find((_package: IPackageRegistryEntry) => {
                return _package.name === (this.props.selectedPackage as IPackageRegistryEntry).name && _package.version === (this.props.selectedPackage as IPackageRegistryEntry).version;
            });

            packageStillExists = !!packageToCheck;

            if (!packageStillExists) {
                // User deleted the selected package.
                this.props.onPackageChange(undefined);
            }
        } else {
            if (this.props.deletedSelectedPackage) {
                // If the selected package was deleted whilst in Step Two or Three
                deletedMessage = (
                    <div className='bx--row margin-bottom-03'>
                        <div className='bx--col-lg-10'>
                            <InlineNotification
                            lowContrast={true}
                            kind='warning'
                            hideCloseButton={true}
                            title='The package you selected has been deleted'
                            />
                        </div>
                    </div>
                );
            }

            if (this.props.selectedWorkspace) {

                packageMessage = (
                    <div className='bx--row margin-bottom-03'>
                        <div className='bx--col-lg-10'>
                            <InlineNotification
                                actions={<NotificationActionButton style={{fontWeight: 600}} onClick={this.packageWorkspace}>Package open project</NotificationActionButton>}
                                hideCloseButton={true}
                                kind='info'
                                lowContrast={true}
                                notificationType='inline'
                                role='alert'
                                statusIconDescription='describes the status icon'
                                subtitle='before it can be used.'
                                title='The contract must be packaged'
                            />
                        </div>
                    </div>
                );

            }
        }

        return(
            <>
                <div className={'bx--row' + (packageMessage || deletedMessage ? '' : ' margin-bottom-10')}>
                    <div className='bx--col'>
                        <div style={{width: 384}}>
                            <Dropdown
                                ariaLabel='dropdown'
                                id='package-select'
                                invalidText='A valid value is required'
                                items={items}
                                label={this.unselectedPackageLabel}
                                titleText='Choose a smart contract to deploy'
                                type='default'
                                selectedItem={this.props.selectedPackage && packageStillExists  ? this.formatPackageEntry(this.props.selectedPackage) : (this.props.selectedWorkspace ? this.formatWorkspaceEntry(this.props.selectedWorkspace) :  this.unselectedPackageLabel)}
                                onChange={this.selectPackage}
                            />
                        </div>
                    </div>
                </div>
                {deletedMessage}
                {packageMessage}
                <div className='bx--row'>
                    <div className='bx--col-lg-10'>
                        <Accordion>
                            <AccordionItem title={'How does Fabric v2.X smart contract deployment work?'}>
                                <div className='bx--row margin-top-05 margin-bottom-09'>
                                    <p>This deployment flow simplifies the steps below and you will only need to know these steps if you intend on using the operator console to operate a Blockchain network. If you do wish to operate a network in the future, or just want to know about how the smart contract deployment works in Fabric v2.0, then continue reading.</p>
                                </div>
                                <div className='bx--row margin-bottom-09'>
                                    <div className='bx--col margin-right-07 padding-right-05'>
                                        <p className='step-title margin-bottom-05'>Step 1</p>
                                        <h6 className='margin-bottom-05'>Your smart contract is packaged</h6>
                                        <p className='margin-bottom-05'>We need to package the chaincode before it can be installed on our peers.</p>
                                    </div>
                                    <div className='bx--col'>
                                        <ThemedImage darkImg={darkStep1} lightImg={lightStep1} altText='' id='step-1-deploy' className=''/>
                                    </div>
                                </div>
                                <div className='bx--row margin-bottom-09'>
                                    <div className='bx--col'>
                                        <ThemedImage darkImg={darkStep2} lightImg={lightStep2} altText='' id='step-2-deploy'/>
                                    </div>
                                    <div className='bx--col margin-left-07 padding-left-05'>
                                        <p className='step-title margin-bottom-05'>Step 2</p>
                                        <h6 className='margin-bottom-05'>Each of the network's member organizations install the package on their peers</h6>
                                        <p className='margin-bottom-05'>After we package the smart contract, we can install the chaincode on our peers. The chaincode needs to be installed on every peer that will endorse a transaction.</p>
                                    </div>
                                </div>
                                <div className='bx--row margin-bottom-09'>
                                    <div className='bx--col margin-right-07 padding-right-05'>
                                        <div className='title-and-tag margin-bottom-05'>
                                            <span>Step 3</span>
                                            <Tag type='blue' className='fabric-2-tag'>New for Fabric v2.0</Tag>
                                        </div>
                                        <h6 className='margin-bottom-05'>Each organization approves a shared definition of what they will use on the channel</h6>
                                        <p className='margin-bottom-05'>After you install the chaincode package, you need to approve a chaincode definition for your organisation. The definition includes important parameters of chaincode governance such as the name, version, and the chaincode endorsement policy.</p>
                                    </div>
                                    <div className='bx--col'>
                                        <ThemedImage darkImg={darkStep3} lightImg={lightStep3} altText='' id='step-3-deploy'/>
                                    </div>
                                </div>
                                <div className='bx--row margin-bottom-09'>
                                    <div className='bx--col'>
                                        <ThemedImage darkImg={darkStep4} lightImg={lightStep4} altText='' id='step-4-deploy'/>
                                    </div>
                                    <div className='bx--col margin-left-07 padding-left-05'>
                                        <div className='title-and-tag margin-bottom-05'>
                                            <span>Step 4</span>
                                            <Tag type='blue' className='fabric-2-tag'>New for Fabric v2.0</Tag>
                                        </div>
                                        <h6 className='margin-bottom-05'>The definition is then committed: a transaction endorsed by the channel members</h6>
                                        <p className='margin-bottom-05'>After a sufficient number of organizations have approved a chaincode definition, one organization can commit the chaincode definition to the channel. If a majority of channel members have approved the definition, the commit transaction will be successful and the parameters agreed to in the chaincode definition will be implemented on the channel.</p>
                                    </div>
                                </div>
                                <div className='bx--row'>
                                    <p>To learn more, visit the <Link href='https://hyperledger-fabric.readthedocs.io/en/release-2.0/chaincode_lifecycle.html' id='deploy-step-1-link'>Fabric chaincode lifecycle</Link> section in the Hyperledger Fabric docs.</p>
                                </div>
                            </AccordionItem>
                        </Accordion>
                    </div>
                </div>
            </>
        );
    }
}

export default DeployStepOne;
