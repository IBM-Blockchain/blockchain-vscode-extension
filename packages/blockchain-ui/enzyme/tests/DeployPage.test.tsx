import React from 'react';
import renderer from 'react-test-renderer';
import chai from 'chai';
import sinon from 'sinon';
import sinonChai from 'sinon-chai';

import DeployPage from '../../src/components/pages/DeployPage/DeployPage';
import IPackageRegistryEntry from '../../src/interfaces/IPackageRegistryEntry';
import { ReactWrapper, mount, shallow, ShallowWrapper } from 'enzyme';
import DeployStepOne from '../../src/components/elements/DeploySteps/DeployStepOne/DeployStepOne';
import DeployStepTwo from '../../src/components/elements/DeploySteps/DeployStepTwo/DeployStepTwo';
import DeployStepThree from '../../src/components/elements/DeploySteps/DeployStepThree/DeployStepThree';
import Utils from '../../src/Utils';

chai.should();
chai.use(sinonChai);

describe('DeployPage component', () => {
    let mySandBox: sinon.SinonSandbox;

    let handleProgressChangeStub: sinon.SinonStub;
    let handlePackageChangeStub: sinon.SinonStub;
    let handleDefinitionNameChange: sinon.SinonStub;
    let handleDefinitionVersionChange: sinon.SinonStub;
    let postToVscodeStub: sinon.SinonStub;

    const packageOne: IPackageRegistryEntry = {name: 'mycontract', version: '0.0.1', path: '/package/one', sizeKB: 9000};
    const packageTwo: IPackageRegistryEntry = {name: 'othercontract', version: '0.0.2', path: '/package/two', sizeKB: 12000};
    const deployData: {channelName: string, environmentName: string, packageEntries: IPackageRegistryEntry[]} = {channelName: 'mychannel', environmentName: 'myEnvironment', packageEntries: [packageOne, packageTwo]};

    beforeEach(async () => {
        mySandBox = sinon.createSandbox();
    });

    afterEach(async () => {
        mySandBox.restore();
    });

    describe('render', () => {
        it('should render the expected snapshot', () => {
            const component: any = renderer
                .create(<DeployPage deployData={deployData} />)
                .toJSON();
            expect(component).toMatchSnapshot();
        });

        it('should load step one component', () => {
            const component: ShallowWrapper<DeployPage> = shallow(<DeployPage deployData={deployData} />);

            const doesStepOneExist: boolean = component.exists(DeployStepOne);
            doesStepOneExist.should.equal(true);

            const doesStepTwoExist: boolean = component.exists(DeployStepTwo);
            doesStepTwoExist.should.equal(false);

            const doesStepThreeExist: boolean = component.exists(DeployStepThree);
            doesStepThreeExist.should.equal(false);
        });

        it('should load step two component', () => {
            const component: ShallowWrapper<DeployPage> = shallow(<DeployPage deployData={deployData} />);

            component.setState({progressIndex: 1});

            const doesStepOneExist: boolean = component.exists(DeployStepOne);
            doesStepOneExist.should.equal(false);

            const doesStepTwoExist: boolean = component.exists(DeployStepTwo);
            doesStepTwoExist.should.equal(true);

            const doesStepThreeExist: boolean = component.exists(DeployStepThree);
            doesStepThreeExist.should.equal(false);
        });

        it('should load step three component', () => {
            const component: ShallowWrapper<DeployPage> = shallow(<DeployPage deployData={deployData} />);

            component.setState({progressIndex: 2});

            const doesStepOneExist: boolean = component.exists(DeployStepOne);
            doesStepOneExist.should.equal(false);

            const doesStepTwoExist: boolean = component.exists(DeployStepTwo);
            doesStepTwoExist.should.equal(false);

            const doesStepThreeExist: boolean = component.exists(DeployStepThree);
            doesStepThreeExist.should.equal(true);
        });
    });

    describe('handleProgressChange', () => {
        it('should update progress state', () => {

            const component: ReactWrapper<DeployPage> = mount(<DeployPage deployData={deployData} />);
            const instance: DeployPage = component.instance() as DeployPage;

            handleProgressChangeStub = mySandBox.stub(instance, 'setState').resolves();

            instance.handleProgressChange(1);

            handleProgressChangeStub.should.have.been.calledOnceWithExactly({progressIndex: 1});

        });

        it('should reset Next button when going back from Step 2 to Step 1 with errors', () => {
            const component: ReactWrapper<DeployPage> = mount(<DeployPage deployData={deployData} />);
            component.setState({selectedPackage: packageOne});
            const instance: DeployPage = component.instance() as DeployPage;

            handleProgressChangeStub = mySandBox.stub(instance, 'setState').resolves();

            instance.handleProgressChange(0);

            handleProgressChangeStub.should.have.been.calledOnceWithExactly({progressIndex: 0, disableNext: false});
        });
    });

    describe('handlePackageChange', () => {
        it('should update package selected and reset definition name and version (so defaults get used)', () => {

            const component: ReactWrapper<DeployPage> = mount(<DeployPage deployData={deployData} />);
            const instance: DeployPage = component.instance() as DeployPage;

            handlePackageChangeStub = mySandBox.stub(instance, 'setState').resolves();

            instance.handlePackageChange(packageTwo);

            handlePackageChangeStub.should.have.been.calledOnceWithExactly({selectedPackage: packageTwo, definitionName: packageTwo.name, definitionVersion: packageTwo.version, disableNext: false});

        });
    });

    describe('handleDefinitionNameChange', () => {
        it('should update definition name', () => {

            const component: ReactWrapper<DeployPage> = mount(<DeployPage deployData={deployData} />);
            const instance: DeployPage = component.instance() as DeployPage;

            handleDefinitionNameChange = mySandBox.stub(instance, 'setState').resolves();

            instance.handleDefinitionNameChange('newName');

            handleDefinitionNameChange.should.have.been.calledOnceWithExactly({definitionName: 'newName', disableNext: false});

        });
    });

    describe('handleDefinitionVersionChange', () => {
        it('should update definition version', () => {

            const component: ReactWrapper<DeployPage> = mount(<DeployPage deployData={deployData} />);
            const instance: DeployPage = component.instance() as DeployPage;

            handleDefinitionVersionChange = mySandBox.stub(instance, 'setState').resolves();

            instance.handleDefinitionVersionChange('0.0.3');

            handleDefinitionVersionChange.should.have.been.calledOnceWithExactly({definitionVersion: '0.0.3', disableNext: false});

        });
    });

    describe('handleDeploy', () => {
        it('should send deploy message', () => {
            postToVscodeStub = mySandBox.stub(Utils, 'postToVSCode').returns(undefined);

            const component: ReactWrapper<DeployPage> = mount(<DeployPage deployData={deployData} />);
            const instance: DeployPage = component.instance() as DeployPage;

            instance.setState({
                environmentName: 'myEnvironment',
                channelName: 'myChannel',
                selectedPackage: packageTwo,
                definitionName: packageTwo.name,
                definitionVersion: packageTwo.version,
                commitSmartContract: undefined
            });

            instance.handleDeploy();

            postToVscodeStub.should.have.been.calledOnceWithExactly({
                command: 'deploy',
                data: {
                    environmentName: 'myEnvironment',
                    channelName: 'myChannel',
                    selectedPackage: packageTwo,
                    definitionName: packageTwo.name,
                    definitionVersion: packageTwo.version,
                    commitSmartContract: undefined
                }
            });
        });
    });

});
