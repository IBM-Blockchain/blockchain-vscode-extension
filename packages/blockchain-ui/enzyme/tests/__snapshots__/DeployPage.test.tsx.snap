// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeployPage component render should render the expected snapshot 1`] = `
<div
  className="bx--grid deploy-page-container"
>
  <div
    className="bx--row"
  >
    <div
      className="bx--col-lg-10"
    >
      <div
        className="heading-combo-container"
      >
        <h3>
          Deploy smart contract
        </h3>
        <p>
          Deploying to mychannel in myEnvironment
        </p>
      </div>
    </div>
    <div
      className="bx--col-lg-4"
    >
      <div
        id="buttonList"
      >
        <button
          className="bx--btn bx--btn--primary bx--btn--disabled"
          disabled={true}
          onClick={[Function]}
          tabIndex={0}
          type="button"
        >
          Next
        </button>
      </div>
    </div>
    <div
      className="bx--col-lg-2"
    />
  </div>
  <div
    className="bx--row progress-row"
  >
    <div
      className="bx--col-lg-10"
    >
      <ul
        className="bx--progress"
      >
        <li
          className="bx--progress-step bx--progress-step--current"
        >
          <button
            className="bx--progress-step-button bx--progress-step-button--unclickable"
            onKeyDown={[Function]}
            tabIndex={-1}
          >
            <span
              className="bx--assistive-text"
            >
              Current
            </span>
            <svg
              aria-hidden={true}
              focusable="false"
              height={16}
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 32 32"
              width={16}
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="16"
                cy="16"
                r="10"
              />
              <path
                d="M16,30A14,14,0,1,1,30,16,14.0158,14.0158,0,0,1,16,30ZM16,4A12,12,0,1,0,28,16,12.0137,12.0137,0,0,0,16,4Z"
              />
              <title>
                
              </title>
            </svg>
            <p
              className="bx--progress-label"
            >
              Step 1
            </p>
            <p
              className="bx--progress-optional"
            >
              Choose smart contract
            </p>
            <span
              className="bx--progress-line"
            />
          </button>
        </li>
        <li
          className="bx--progress-step bx--progress-step--incomplete"
        >
          <button
            className="bx--progress-step-button bx--progress-step-button--unclickable"
            onKeyDown={[Function]}
            tabIndex={-1}
          >
            <span
              className="bx--assistive-text"
            >
              Incomplete
            </span>
            <svg
              aria-hidden={true}
              focusable="false"
              height={16}
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 32 32"
              width={16}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M16,2A14,14,0,1,0,30,16,14,14,0,0,0,16,2Zm0,26A12,12,0,1,1,28,16,12,12,0,0,1,16,28Z"
              />
              <title>
                
              </title>
            </svg>
            <p
              className="bx--progress-label"
            >
              Step 2
            </p>
            <p
              className="bx--progress-optional"
            >
              Create definition
            </p>
            <span
              className="bx--progress-line"
            />
          </button>
        </li>
        <li
          className="bx--progress-step bx--progress-step--incomplete"
        >
          <button
            className="bx--progress-step-button bx--progress-step-button--unclickable"
            onKeyDown={[Function]}
            tabIndex={-1}
          >
            <span
              className="bx--assistive-text"
            >
              Incomplete
            </span>
            <svg
              aria-hidden={true}
              focusable="false"
              height={16}
              preserveAspectRatio="xMidYMid meet"
              viewBox="0 0 32 32"
              width={16}
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M16,2A14,14,0,1,0,30,16,14,14,0,0,0,16,2Zm0,26A12,12,0,1,1,28,16,12,12,0,0,1,16,28Z"
              />
              <title>
                
              </title>
            </svg>
            <p
              className="bx--progress-label"
            >
              Step 3
            </p>
            <p
              className="bx--progress-optional"
            >
              Deploy
            </p>
            <span
              className="bx--progress-line"
            />
          </button>
        </li>
      </ul>
    </div>
  </div>
  <div
    className="bx--row margin-bottom-10"
  >
    <div
      className="bx--col"
    >
      <div
        style={
          Object {
            "width": 384,
          }
        }
      >
        <div
          className="bx--dropdown__wrapper bx--list-box__wrapper"
        >
          <span
            className="bx--label"
          >
            Choose a smart contract to deploy
          </span>
          <div
            aria-label="dropdown"
            className="bx--dropdown bx--list-box"
            id="package-select"
            onClick={[Function]}
            onKeyDown={[Function]}
          >
            <div
              aria-controls={null}
              aria-disabled={false}
              aria-expanded={false}
              aria-haspopup={true}
              aria-labelledby="dropdown-label-package-select dropdown-field-label-package-select"
              aria-owns={null}
              className="bx--list-box__field"
              data-toggle={true}
              onBlur={[Function]}
              onClick={[Function]}
              onKeyDown={[Function]}
              role="button"
              tabIndex="0"
              type="button"
            >
              <span
                className="bx--list-box__label"
                htmlFor="downshift-0-input"
                id="dropdown-field-label-package-select"
              >
                Select smart contract
              </span>
              <div
                className="bx--list-box__menu-icon"
              >
                <svg
                  aria-label="Open menu"
                  focusable="false"
                  height={16}
                  name="chevron--down"
                  preserveAspectRatio="xMidYMid meet"
                  role="img"
                  viewBox="0 0 16 16"
                  width={16}
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <polygon
                    points="8,11 3,6 3.7,5.3 8,9.6 12.3,5.3 13,6"
                  />
                  <title>
                    Open menu
                  </title>
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="bx--row"
  >
    <div
      className="bx--col-lg-10"
    >
      <ul
        className="bx--accordion bx--accordion--end"
      >
        <li
          className="bx--accordion__item"
          onAnimationEnd={[Function]}
        >
          <button
            aria-expanded={false}
            className="bx--accordion__heading"
            onClick={[Function]}
            onKeyDown={[Function]}
            title="Expand/Collapse"
            type="button"
          >
            <svg
              aria-label="Expand/Collapse"
              className="bx--accordion__arrow"
              focusable="false"
              height={16}
              preserveAspectRatio="xMidYMid meet"
              role="img"
              viewBox="0 0 16 16"
              width={16}
              xmlns="http://www.w3.org/2000/svg"
            >
              <polygon
                points="11,8 6,13 5.3,12.3 9.6,8 5.3,3.7 6,3"
              />
            </svg>
            <div
              className="bx--accordion__title"
            >
              How does Fabric v2.X smart contract deployment work?
            </div>
          </button>
          <div
            className="bx--accordion__content"
          >
            <div
              className="bx--row margin-top-05 margin-bottom-09"
            >
              <p>
                This deployment flow simplifies the steps below and you will only need to know these steps if you intend on using the operator console to operate a Blockchain network. If you do wish to operate a network in the future, or just want to know about how the smart contract deployment works in Fabric v2.0, then continue reading.
              </p>
            </div>
            <div
              className="bx--row margin-bottom-09"
            >
              <div
                className="bx--col margin-right-07 padding-right-05"
              >
                <p
                  className="step-title margin-bottom-05"
                >
                  Step 1
                </p>
                <h6
                  className="margin-bottom-05"
                >
                  Your smart contract is packaged
                </h6>
                <p
                  className="margin-bottom-05"
                >
                  We need to package the chaincode before it can be installed on our peers.
                </p>
              </div>
              <div
                className="bx--col"
              >
                <div
                  id="step-1-deploy-themed-image"
                >
                  <img
                    alt=""
                    className=""
                    src="test-file-stub"
                  />
                  <img
                    alt=""
                    className=""
                    src="test-file-stub"
                  />
                </div>
              </div>
            </div>
            <div
              className="bx--row margin-bottom-09"
            >
              <div
                className="bx--col"
              >
                <div
                  id="step-2-deploy-themed-image"
                >
                  <img
                    alt=""
                    className=""
                    src="test-file-stub"
                  />
                  <img
                    alt=""
                    className=""
                    src="test-file-stub"
                  />
                </div>
              </div>
              <div
                className="bx--col margin-left-07 padding-left-05"
              >
                <p
                  className="step-title margin-bottom-05"
                >
                  Step 2
                </p>
                <h6
                  className="margin-bottom-05"
                >
                  Each of the network's member organizations install the package on their peers
                </h6>
                <p
                  className="margin-bottom-05"
                >
                  After we package the smart contract, we can install the chaincode on our peers. The chaincode needs to be installed on every peer that will endorse a transaction.
                </p>
              </div>
            </div>
            <div
              className="bx--row margin-bottom-09"
            >
              <div
                className="bx--col margin-right-07 padding-right-05"
              >
                <div
                  className="title-and-tag margin-bottom-05"
                >
                  <span>
                    Step 3
                  </span>
                  <span
                    className="bx--tag bx--tag--blue fabric-2-tag"
                  >
                    New for Fabric v2.0
                  </span>
                </div>
                <h6
                  className="margin-bottom-05"
                >
                  Each organization approves a shared definition of what they will use on the channel
                </h6>
                <p
                  className="margin-bottom-05"
                >
                  After you install the chaincode package, you need to approve a chaincode definition for your organisation. The definition includes important parameters of chaincode governance such as the name, version, and the chaincode endorsement policy.
                </p>
              </div>
              <div
                className="bx--col"
              >
                <div
                  id="step-3-deploy-themed-image"
                >
                  <img
                    alt=""
                    className=""
                    src="test-file-stub"
                  />
                  <img
                    alt=""
                    className=""
                    src="test-file-stub"
                  />
                </div>
              </div>
            </div>
            <div
              className="bx--row margin-bottom-09"
            >
              <div
                className="bx--col"
              >
                <div
                  id="step-4-deploy-themed-image"
                >
                  <img
                    alt=""
                    className=""
                    src="test-file-stub"
                  />
                  <img
                    alt=""
                    className=""
                    src="test-file-stub"
                  />
                </div>
              </div>
              <div
                className="bx--col margin-left-07 padding-left-05"
              >
                <div
                  className="title-and-tag margin-bottom-05"
                >
                  <span>
                    Step 4
                  </span>
                  <span
                    className="bx--tag bx--tag--blue fabric-2-tag"
                  >
                    New for Fabric v2.0
                  </span>
                </div>
                <h6
                  className="margin-bottom-05"
                >
                  The definition is then committed: a transaction endorsed by the channel members
                </h6>
                <p
                  className="margin-bottom-05"
                >
                  After a sufficient number of organizations have approved a chaincode definition, one organization can commit the chaincode definition to the channel. If a majority of channel members have approved the definition, the commit transaction will be successful and the parameters agreed to in the chaincode definition will be implemented on the channel.
                </p>
              </div>
            </div>
            <div
              className="bx--row"
            >
              <p>
                To learn more, visit the 
                <a
                  className="bx--link"
                  href="https://hyperledger-fabric.readthedocs.io/en/release-2.0/chaincode_lifecycle.html"
                  id="deploy-step-1-link"
                >
                  Fabric chaincode lifecycle
                </a>
                 section in the Hyperledger Fabric docs.
              </p>
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;
